crates/base-db/src/lib.rs

```diff
@@ -47,39 +50,202 @@ pub const DEFAULT_FILE_TEXT_LRU_CAP: u16 = 16;
+                vacant.insert(text);
+            }
+        };
+    }
 
+    pub fn set_file_text_with_durability(
+        &self,
+        db: &mut dyn SourceDatabase,
+        file_id: vfs::FileId,
+        text: &str,
+        durability: Durability,
+    ) {
+        let files = Arc::clone(&self.files);
+        match files.entry(file_id) {
+            Entry::Occupied(mut occupied) => {
+                occupied.get_mut().set_text(db).to(Arc::from(text));
+            }
+            Entry::Vacant(vacant) => {
+                let text =
+                    FileText::builder(Arc::from(text), file_id).durability(durability).new(db);
+                vacant.insert(text);
+            }
+        };
+    }
+
+    /// Source root of the file.
+    pub fn source_root(&self, source_root_id: SourceRootId) -> SourceRootInput {
+        let source_root = self
+            .source_roots
+            .get(&source_root_id)
+            .expect("Unable to fetch source root id; this is a bug");
+
+        *source_root
+    }
+
+    pub fn set_source_root_with_durability(
+        &self,
+        db: &mut dyn SourceDatabase,
+        source_root_id: SourceRootId,
+        source_root: Arc<SourceRoot>,
+        durability: Durability,
+    ) {
+        let source_roots = Arc::clone(&self.source_roots);
+        match source_roots.entry(source_root_id) {
+            Entry::Occupied(mut occupied) => {
+                occupied.get_mut().set_source_root(db).to(source_root);
+            }
+            Entry::Vacant(vacant) => {
+                let source_root =
+                    SourceRootInput::builder(source_root).durability(durability).new(db);
+                vacant.insert(source_root);
+            }
+        };
+    }
+
+    pub fn file_source_root(&self, id: vfs::FileId) -> FileSourceRootInput {
+        let file_source_root = self
+            .file_source_roots
+            .get(&id)
+            .expect("Unable to fetch FileSourceRootInput; this is a bug");
+        *file_source_root
+    }
+
+    pub fn set_file_source_root_with_durability(
+        &self,
+        db: &mut dyn SourceDatabase,
+        id: vfs::FileId,
+        source_root_id: SourceRootId,
+        durability: Durability,
+    ) {
+        let file_source_roots = Arc::clone(&self.file_source_roots);
+        // let db = self;
+        match file_source_roots.entry(id) {
+            Entry::Occupied(mut occupied) => {
+                occupied.get_mut().set_source_root_id(db).to(source_root_id);
+            }
+            Entry::Vacant(vacant) => {
+                let file_source_root =
+                    FileSourceRootInput::builder(source_root_id).durability(durability).new(db);
+                vacant.insert(file_source_root);
+            }
+        };
+    }
+}
+
+#[salsa::interned(no_lifetime)]
+pub struct EditionedFileId {
+    pub file_id: vfs::FileId,
+    pub editioned_file_id: span::EditionedFileId,
```

Comment:
Oh ye that would do it, this definition is recursive :) That should be just an edition

Reply:
Hm okay i see, thats because the source database is already salsafied so we need a different input here. That is quite annoying. Either way the definition of this should be
```rust
pub struct EditionedFileId {
    pub file_id: vfs::FileId,
    pub edition: span::Edition
}
```
or

```rust
pub struct EditionedFileId {
    pub file_id: span::EditionedFileId,
}
```

Right now it encodes the file id twice resulting in very weird transformation code. We also should introduce a help func that transforms between the span and base-db versions

Reply:
I think I'd prefer

```rust
pub struct EditionedFileId {
    pub file_id: span::EditionedFileId,
}
```

...because I'm pretty sure `span::EditionedFileId` is used in crates without a Salsa dep: I remember this was the case about 3 months ago and stumbling over it. Would rather just have the ID here than reconstruct it each time.

I'll add some methods which make deconstructing the value nicer.

Reply:
done.